src/twisted/application/runner/test/test_pidfile.py

```diff
@@ -0,0 +1,291 @@
+"""
+
+import errno
+from os import getpid, name as SYSTEM_NAME
+from io import BytesIO
+
+from twisted.python.filepath import FilePath
+
+from ...runner import _pidfile
+from .._pidfile import PIDFile, AlreadyRunningError
+
+import twisted.trial.unittest
+from twisted.trial.unittest import SkipTest
+
+
+
+class PIDFileTests(twisted.trial.unittest.TestCase):
+    """
+    Tests for L{PIDFile}.
+    """
+
+    def test_formatWithPID(self):
+        """
+        L{PIDFile.format} returns the expected format when given a PID.
+        """
+        self.assertEqual(PIDFile.format(pid=1337), b"1337\n")
+
+
+    def test_formatWithoutPID(self):
+        """
+        L{PIDFile.format} returns the expected format when not given a PID.
+        """
+        self.assertEqual(PIDFile.format(), b"")
+
+
+    def test_readWithPID(self):
+        """
+        L{PIDFile.read} returns the PID from the given file path.
+        """
+        pid = 1337
+
+        pidFile = PIDFile(DummyFilePath(PIDFile.format(pid=pid)))
+
+        self.assertEqual(pid, pidFile.read())
+
+
+    def test_readWithoutPID(self):
+        """
+        L{PIDFile.read} raises ValueError when given an empty file path.
+        """
+        pidFile = PIDFile(DummyFilePath(b""))
+
+        self.assertRaises(ValueError, pidFile.read)
+
+
+    def test_readWithBogusPID(self):
+        """
+        L{PIDFile.read} raises ValueError when given an invalid file path.
+        """
+        pidFile = PIDFile(DummyFilePath(b"not a pid"))
+
+        self.assertRaises(ValueError, pidFile.read)
+
+
+    def test_writeDefault(self):
+        """
+        L{PIDFile.write} with no C{pid} argument stores the PID for the current
+            process.
+        """
+        pidFile = PIDFile(DummyFilePath())
+        pidFile.write()
+
+        self.assertEqual(pidFile.read(), getpid())
+
+
+    def test_writePID(self):
+        """
+        L{PIDFile.write} stores the given PID.
+        """
+        pid = 1995
+
+        pidFile = PIDFile(DummyFilePath())
+        pidFile.write(pid)
+
+        self.assertEqual(pidFile.read(), pid)
+
+
+    def test_writePIDInvalid(self):
+        """
+        L{PIDFile.write} raises ValueError when given an invalid PID.
+        """
+        pidFile = PIDFile(DummyFilePath())
+
+        self.assertRaises(ValueError, pidFile.write, u"burp")
+
+
+    def test_remove(self):
+        """
+        L{PIDFile.remove} removes the PID file.
+        """
+        pidFile = PIDFile(DummyFilePath(b""))
+        self.assertTrue(pidFile.filePath.exists())
+
+        pidFile.remove()
+        self.assertFalse(pidFile.filePath.exists())
+
+
+    def test_isRunningDoesExist(self):
+        """
+        L{PIDFile.isRunning} returns true for a process that does exist.
+        """
+        pidFile = PIDFile(DummyFilePath())
+        pidFile.write(1337)
+
+        def kill(pid, signal):
+            return  # Don't actually kill anything
+
+        self.patch(_pidfile, "kill", kill)
```

Comment:
I don't know how I feel about this, but I'm not sure I have a better alternative.